File: chapter06/modules_and_functions_6.ex
defmodule Chop do
  def guess(actual, range = min..max) do
    helper(actual, range, average(min, max))
  end

  defp average(min, max), do: div(min + max, 2)

  defp helper(actual, min.._, expected)
  when actual < expected do
    IO.puts "Is it #{inspect expected}"
    helper(actual, min..(expected - 1), average(min, expected - 1))
  end

  defp helper(actual, _..max, expected)
  when actual > expected do
    IO.puts "Is it #{inspect expected}"
    helper(actual, (expected + 1)..max, average(expected + 1, max))
  end

  defp helper(actual, _, expected) when actual == expected, do: expected
end

File: chapter06/modules_and_functions_6.clj
(ns modules-and-functions-6)

(defn- average [min max]
  (quot (+ min max) 2))

(defn- helper [actual min max expected]
  (cond
    (< actual expected)
    (do (println "Is it" expected)
        (recur actual min (dec expected) (average min (dec expected))))

    (> actual expected)
    (do (println "Is it" expected)
        (recur actual (inc expected) max (average (inc expected) max)))

    :else expected))

(defn guess [actual min max]
  (helper actual min max (average min max)))
